File: pyspider/core/proxy_pool.py
"""
代理池管理器 - 支持多代理轮换、健康检查和自动故障转移
"""

import time
import threading
import requests
from typing import Optional, List, Dict
from dataclasses import dataclass
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class ProxyStatus(Enum):
    UNKNOWN = "unknown"
    ALIVE = "alive"
    DEAD = "dead"


@dataclass
class Proxy:
    """代理服务器"""

    host: str
    port: int
    protocol: str = "http"  # http, https, socks5
    username: Optional[str] = None
    password: Optional[str] = None
    status: ProxyStatus = ProxyStatus.UNKNOWN
    success_count: int = 0
    failure_count: int = 0
    last_checked: Optional[float] = None
    response_time: float = 0.0

    @property
    def url(self) -> str:
        """获取代理 URL"""
        if self.username and self.password:
            return f"{self.protocol}://{self.username}:{self.password}@{self.host}:{self.port}"
        return f"{self.protocol}://{self.host}:{self.port}"

    @property
    def proxies_dict(self) -> Dict[str, str]:
        """获取 requests 格式的代理字典"""
        return {"http": self.url, "https": self.url}

    @property
    def success_rate(self) -> float:
        """计算成功率"""
        total = self.success_count + self.failure_count
        return self.success_count / total if total > 0 else 0.0


class ProxyPool:
    """代理池管理器"""

    def __init__(
        self,
        check_url: str = "https://httpbin.org/ip",
        check_interval: int = 300,
        max_failures: int = 3,
    ):
        self._proxies: List[Proxy] = []
        self._lock = threading.RLock()
        self._current_index = 0
        self._check_url = check_url
        self._check_interval = check_interval
        self._max_failures = max_failures
        self._health_check_running = False
        self._health_check_thread: Optional[threading.Thread] = None

    def add_proxy(self, proxy: Proxy) -> None:
        """添加代理"""
        with self._lock:
            self._proxies.append(proxy)
            logger.info(f"代理已添加: {proxy.url}")

    def add_proxies(self, proxies: List[Proxy]) -> None:
        """批量添加代理"""
        for proxy in proxies:
            self.add_proxy(proxy)

    def add_proxy_from_string(self, proxy_str: str) -> None:
        """
        从字符串添加代理
        支持格式:
        - http://host:port
        - http://user:pass@host:port
        - host:port
        """
        import re

        # 修复: 使用非贪婪匹配支持密码中的特殊字符
        pattern = r"^(?:(\w+)://)?(?:(.+?):(.+?)@)?([\w.-]+):(\d+)$"
        match = re.match(pattern, proxy_str)
        if match:
            protocol = match.group(1) or "http"
            username = match.group(2)
            password = match.group(3)
            host = match.group(4)
            port = int(match.group(5))
            self.add_proxy(
                Proxy(
                    host=host,
                    port=port,
                    protocol=protocol,
                    username=username,
                    password=password,
                )
            )
        else:
            logger.warning(f"代理格式无效: {proxy_str}")

    def get_proxy(self) -> Optional[Proxy]:
        """获取下一个可用代理(轮询+健康状态)"""
        with self._lock:
            if not self._proxies:
                return None

            # 尝试最多 len(proxies) 次找到可用代理
            for _ in range(len(self._proxies)):
                proxy = self._proxies[self._current_index % len(self._proxies)]
                self._current_index += 1

                if (
                    proxy.status == ProxyStatus.ALIVE
                    and proxy.failure_count < self._max_failures
                ):
                    return proxy

            # 如果没有健康代理,返回第一个未知状态的代理
            for proxy in self._proxies:
                if proxy.status == ProxyStatus.UNKNOWN:
                    return proxy

            # 所有代理都不可用,返回 None
            return None

    def record_success(self, proxy: Proxy) -> None:
        """记录成功"""
        with self._lock:
            proxy.success_count += 1
            proxy.status = ProxyStatus.ALIVE
            proxy.last_checked = time.time()

    def record_failure(self, proxy: Proxy) -> None:
        """记录失败"""
        with self._lock:
            proxy.failure_count += 1
            proxy.last_checked = time.time()
            if proxy.failure_count >= self._max_failures:
                proxy.status = ProxyStatus.DEAD
                logger.warning(f"代理已标记为死亡: {proxy.url}")

    def check_proxy(self, proxy: Proxy, timeout: int = 10) -> bool:
        """检查代理是否可用"""
        try:
            start = time.time()
            resp = requests.get(
                self._check_url, proxies=proxy.proxies_dict, timeout=timeout
            )
            proxy.response_time = time.time() - start
            if resp.status_code == 200:
                proxy.failure_count = 0  # 修复: 恢复时重置失败计数
                self.record_success(proxy)
                return True
            else:
                self.record_failure(proxy)
                return False
        except Exception as e:
            logger.debug(f"代理检查失败 {proxy.url}: {e}")
            self.record_failure(proxy)
            return False

    def start_health_check(self) -> None:
        """启动健康检查后台线程"""
        if self._health_check_running:
            return

        self._health_check_running = True
        self._health_check_thread = threading.Thread(
            target=self._health_check_loop, daemon=True
        )
        self._health_check_thread.start()
        logger.info("代理健康检查已启动")

    def stop_health_check(self) -> None:
        """停止健康检查"""
        self._health_check_running = False
        if self._health_check_thread:
            self._health_check_thread.join(timeout=5)
        logger.info("代理健康检查已停止")

    def _health_check_loop(self) -> None:
        """健康检查循环"""
        while self._health_check_running:
            with self._lock:
                proxies_to_check = list(self._proxies)

            for proxy in proxies_to_check:
                if proxy.status == ProxyStatus.DEAD:
                    # 定期检查死亡代理是否可以恢复
                    if (
                        proxy.last_checked
                        and (time.time() - proxy.last_checked)
                        > self._check_interval * 2
                    ):
                        self.check_proxy(proxy)
                else:
                    self.check_proxy(proxy)

            time.sleep(self._check_interval)

    def check_all(self, timeout: int = 10) -> Dict[str, bool]:
        """检查所有代理"""
        results = {}
        with self._lock:
            for proxy in self._proxies:
                results[proxy.url] = self.check_proxy(proxy, timeout)
        return results

    @property
    def size(self) -> int:
        """代理池大小"""
        return len(self._proxies)

    @property
    def alive_count(self) -> int:
        """可用代理数量"""
        return sum(1 for p in self._proxies if p.status == ProxyStatus.ALIVE)

    @property
    def dead_count(self) -> int:
        """死亡代理数量"""
        return sum(1 for p in self._proxies if p.status == ProxyStatus.DEAD)

    def stats(self) -> Dict:
        """代理池统计"""
        return {
            "total": self.size,
            "alive": self.alive_count,
            "dead": self.dead_count,
            "unknown": sum(1 for p in self._proxies if p.status == ProxyStatus.UNKNOWN),
        }

File: javaspider/src/main/java/com/javaspider/core/ProxyPool.java
package com.javaspider.core;

import java.util.*;
import java.util.concurrent.*;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 代理池管理器 - 支持多代理轮换、健康检查和自动故障转移
 */
public class ProxyPool {
    private static final Logger logger = Logger.getLogger(ProxyPool.class.getName());
    
    public enum ProxyStatus {
        UNKNOWN, ALIVE, DEAD
    }
    
    public static class Proxy {
        public String host;
        public int port;
        public String protocol = "http";
        public String username;
        public String password;
        public ProxyStatus status = ProxyStatus.UNKNOWN;
        public int successCount = 0;
        public int failureCount = 0;
        public long lastChecked = 0;
        public long responseTime = 0;
        
        public Proxy(String host, int port) {
            this.host = host;
            this.port = port;
        }
        
        public Proxy(String host, int port, String protocol) {
            this.host = host;
            this.port = port;
            this.protocol = protocol;
        }
        
        public String getUrl() {
            if (username != null && !username.isEmpty()) {
                return String.format("%s://%s:%s@%s:%d", protocol, username, password, host, port);
            }
            return String.format("%s://%s:%d", protocol, host, port);
        }
        
        public float getSuccessRate() {
            int total = successCount + failureCount;
            return total > 0 ? (float) successCount / total : 0.0f;
        }
    }
    
    private final List<Proxy> proxies = Collections.synchronizedList(new ArrayList<>());
    private volatile int currentIndex = 0;
    private final String checkUrl;
    private final long checkIntervalMs;
    private final int maxFailures;
    private volatile boolean running = false;
    private ExecutorService healthCheckExecutor;
    
    public ProxyPool() {
        this("https://httpbin.org/ip", 300_000, 3);
    }
    
    public ProxyPool(String checkUrl, long checkIntervalMs, int maxFailures) {
        this.checkUrl = checkUrl;
        this.checkIntervalMs = checkIntervalMs;
        this.maxFailures = maxFailures;
    }
    
    public void addProxy(Proxy proxy) {
        proxies.add(proxy);
        logger.info("Proxy added: " + proxy.getUrl());
    }
    
    public void addProxyFromString(String proxyStr) {
        // 支持格式: http://host:port, http://user:pass@host:port, host:port
        Pattern pattern = Pattern.compile("^(?:(\\w+)://)?(?:(\\w+):(\\w+)@)?([\\w.-]+):(\\d+)$");
        Matcher matcher = pattern.matcher(proxyStr);
        if (matcher.matches()) {
            String protocol = matcher.group(1) != null ? matcher.group(1) : "http";
            String username = matcher.group(2);
            String password = matcher.group(3);
            String host = matcher.group(4);
            int port = Integer.parseInt(matcher.group(5));
            
            Proxy proxy = new Proxy(host, port, protocol);
            proxy.username = username;
            proxy.password = password;
            addProxy(proxy);
        } else {
            logger.warning("Invalid proxy format: " + proxyStr);
        }
    }
    
    public synchronized Proxy getProxy() {
        if (proxies.isEmpty()) {
            return null;
        }
        
        // 尝试最多 len(proxies) 次找到可用代理
        for (int i = 0; i < proxies.size(); i++) {
            int idx = (currentIndex + i) % proxies.size();
            Proxy proxy = proxies.get(idx);
            if (proxy.status == ProxyStatus.ALIVE && proxy.failureCount < maxFailures) {
                currentIndex = (idx + 1) % proxies.size();
                return proxy;
            }
        }
        
        // 返回未知状态代理
        for (Proxy proxy : proxies) {
            if (proxy.status == ProxyStatus.UNKNOWN) {
                return proxy;
            }
        }
        
        return null; // 所有代理都不可用
    }
    
    public void recordSuccess(Proxy proxy) {
        if (proxy == null) return;
        proxy.successCount++;
        proxy.status = ProxyStatus.ALIVE;
        proxy.lastChecked = System.currentTimeMillis();
    }
    
    public void recordFailure(Proxy proxy) {
        if (proxy == null) return;
        proxy.failureCount++;
        proxy.lastChecked = System.currentTimeMillis();
        if (proxy.failureCount >= maxFailures) {
            proxy.status = ProxyStatus.DEAD;
            logger.warning("Proxy marked as dead: " + proxy.getUrl());
        }
    }
    
    public void startHealthCheck() {
        if (running) return;
        running = true;
        healthCheckExecutor = Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "proxy-health-check");
            t.setDaemon(true);
            return t;
        });
        healthCheckExecutor.submit(this::healthCheckLoop);
        logger.info("Proxy health check started");
    }
    
    public void stopHealthCheck() {
        running = false;
        if (healthCheckExecutor != null) {
            healthCheckExecutor.shutdownNow();
        }
        logger.info("Proxy health check stopped");
    }
    
    private void healthCheckLoop() {
        while (running) {
            List<Proxy> copy = new ArrayList<>(proxies);
            for (Proxy proxy : copy) {
                if (!running) break;
                if (proxy.status == ProxyStatus.DEAD) {
                    if (System.currentTimeMillis() - proxy.lastChecked > checkIntervalMs * 2) {
                        checkProxy(proxy);
                    }
                } else {
                    checkProxy(proxy);
                }
            }
            try {
                Thread.sleep(checkIntervalMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }
    
    private void checkProxy(Proxy proxy) {
        if (proxy == null) return;
        
        try {
            long startTime = System.currentTimeMillis();
            
            // 使用 HttpURLConnection 进行健康检查
            java.net.URL url = new java.net.URL(checkUrl);
            java.net.Proxy proxyConn = new java.net.Proxy(
                java.net.Proxy.Type.HTTP,
                new java.net.InetSocketAddress(proxy.host, proxy.port)
            );
            
            java.net.HttpURLConnection conn = (java.net.HttpURLConnection) url.openConnection(proxyConn);
            conn.setConnectTimeout(5000);
            conn.setReadTimeout(5000);
            conn.setRequestMethod("GET");
            
            int responseCode = conn.getResponseCode();
            long responseTime = System.currentTimeMillis() - startTime;
            
            conn.disconnect();
            
            if (responseCode == 200) {
                proxy.status = ProxyStatus.ALIVE;
                proxy.responseTime = responseTime;
                proxy.successCount++;
                proxy.lastChecked = System.currentTimeMillis();
            } else {
                proxy.failureCount++;
                if (proxy.failureCount >= maxFailures) {
                    proxy.status = ProxyStatus.DEAD;
                }
            }
        } catch (Exception e) {
            proxy.failureCount++;
            proxy.lastChecked = System.currentTimeMillis();
            if (proxy.failureCount >= maxFailures) {
                proxy.status = ProxyStatus.DEAD;
                logger.warning("Proxy marked as dead: " + proxy.getUrl() + " - " + e.getMessage());
            }
        }
    }
    
    public int size() { return proxies.size(); }
    
    public int aliveCount() {
        return (int) proxies.stream().filter(p -> p.status == ProxyStatus.ALIVE).count();
    }
    
    public int deadCount() {
        return (int) proxies.stream().filter(p -> p.status == ProxyStatus.DEAD).count();
    }
    
    public Map<String, Object> stats() {
        Map<String, Object> s = new HashMap<>();
        s.put("total", size());
        s.put("alive", aliveCount());
        s.put("dead", deadCount());
        s.put("unknown", (int) proxies.stream().filter(p -> p.status == ProxyStatus.UNKNOWN).count());
        return s;
    }
    
    // 常用 User-Agent 池
    public static final String[] USER_AGENTS = {
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.2 Safari/605.1.15",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    };
    
    private static volatile int uaIndex = 0;
    public static synchronized String getNextUserAgent() {
        return USER_AGENTS[(uaIndex++) % USER_AGENTS.length];
    }
}
